src/helpers/ring.rs

```diff
@@ -26,183 +27,270 @@ pub enum HelperAddr {
 
-            Self {
-                input_queue: tx,
-                left: None,
-                right: None,
-                buf,
-            }
-        }
-
-        fn set_left(&mut self, left: Sender<MessageEnvelope>) {
-            self.left = Some(left);
-        }
+            // Take ownership of the receivers - if they've been created already, move them out
+            // otherwise create them.
+            let rx = match table.active_routes.entry(protocol_id) {
+                Entry::Occupied(mut entry) => entry.get_mut().inbound.take().unwrap(),
+                Entry::Vacant(entry) => entry.insert(Route::new()).inbound.take().unwrap(),
+            };
 
-        fn set_right(&mut self, right: Sender<MessageEnvelope>) {
-            self.right = Some(right);
+            TestChannel {
+                protocol_id,
+                table: Arc::clone(&self.routing_table),
+                senders: HashMap::from([
+                    (HelperAddr::Left, self.left.clone()),
+                    (HelperAddr::Right, self.right.clone()),
+                ]),
+                receivers: rx,
+            }
         }
     }
 
     #[async_trait]
-    impl Ring for TestHelper {
-        async fn send<T: Message>(&self, dest: HelperAddr, msg: T) -> Result<(), Error> {
-            assert!(self.left.is_some());
-            assert!(self.right.is_some());
-
+    impl Ring for TestChannel {
+        async fn send<T: Message>(&mut self, dest: HelperAddr, msg: T) -> Result<(), Error> {
             // inside the envelope we store the sender of the message (i.e. source)
             // but this method accepts the destination. To obtain source from destination
             // we invert it - message send to the left helper is originated from helper on the
             // right side.
-            let (target, source) = match dest {
-                HelperAddr::Left => (self.left.as_ref().unwrap(), HelperAddr::Right),
-                HelperAddr::Right => (self.right.as_ref().unwrap(), HelperAddr::Left),
+            let source = match dest {
+                HelperAddr::Left => HelperAddr::Right,
+                HelperAddr::Right => HelperAddr::Left,
             };
 
             let bytes = serde_json::to_vec(&msg).unwrap().into_boxed_slice();
             let envelope = MessageEnvelope {
-                type_id: TypeId::of::<T>(),
                 source,
+                protocol_id: self.protocol_id,
                 payload: bytes,
             };
 
-            target.send(envelope).await.map_err(|e| Error::SendError {
+            let sender = self.senders.get(&dest).ok_or(SendError {
                 dest,
-                inner: Box::new(e) as _,
+                inner: "No sender for this destination".into(),
             })?;
-            Ok(())
+
+            sender.send(envelope).await.map_err(|e| Error::SendError {
+                dest,
+                inner: Box::new(e) as _,
+            })
```

Comment:
Inconsistent use of SendError vs Error::SendError here.

(I can't follow what the .into() is doing for the former.)

Reply:
ack, better to use just `SendError` imo.

Any string can be converted to a boxed error apparently which is convenient in tests: https://doc.rust-lang.org/src/std/error.rs.html#243-283